tlvf: Variable length lists support

tomereli opened the pull request:
![Screenshot from 2019-07-18 00-33-01](https://user-images.githubusercontent.com/13386849/61413354-eb513d80-a8f3-11e9-8b0a-4cb9f4c61fd2.png)
Fixes #8, #116

There are 2 types of variable (dynamic) length lists
in the TLV factory:
- known length
- unknown length

From the programmer point of view, these are almost identical. The main difference is that known length lists have the length variable part of the TLV.
Uknown length lists have to be the last member in the TLV so their length can be calculated
using the TLV length field.

TLVF currently only supports a single dynamic length list, and it has to be the last member in the class
(be placed last on the buffer). The reason for that is that all the members are written directly on the buffer (and the class itself holds pointers to their location on the buffer),
so allocating a new element on a list that's not in the end of the buffer requires relocating all the data that comes after it.

This limitation is currently preventing us from properly implementing some types of messages,
such as M1 & M2 frames, since those frames require multiple dynamic-length strings.

This PR adds full support for variable length lists as follows:
Dynamic length member have an alloc method which is used to allocate room on the buffer for the newly created list member.

This PR adds a `std::copy_n` to the alloc function which moves the remainder of the buffer
by the allocation size, and adds a marker which we use on every member which comes after this list to update its pointer.
For dynamic list members which are of class type, we also update the create method with a similar
`std::copy_n` since the create method adds the class on the buffer with the initial size, before
allocation method can be used.

Next, this PR aligns M1 and M2 TLVs to the WiFi Simple Setup (WSC) v2.0.6 specification as defined in the EasyMesh Multi-AP specification.

Finally, this PR adds README.md to the TLV factory with the design document and user guide.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add .vscode/launch.json

Debugging TLVF is hard, but it is much easier when using a debugger.
Add .vscode/launch.json which sets up debugging tlv factory python
script in vscode editor.

It can also be used to re-generate all autogenerated files (currently in
the framework only) automatically without building which can be useful
to see generated output without going through the whole build.
To do that, simply open tlvf.py in vscode and click F5.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add getBuffRemainingBytes() overload API

Add getBuffRemainingBytes(void *start) API which will be used in
followup commits for adding variable length lists support in TLVF.
The new API simply returns the number of bytes from the start pointer
till the buffer end. This will be used to `memmove` data on allocation
of variable length list.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: fix missing swap for external enums

When a  class contains an enum member which is defined in an external
yaml file, it's type is not identified as ENUM by the tlvf.py
generateObject() method, which is incorrect and causing the script
not to generate swapping logic for this member.
In case where the enum is not a single byte, it is a real bug.

Fix this by adding logic in generateObject() to correctly identify
ENUM members even if they are in an external namespace.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: variable length list support

There are 2 types of variable (dynamic) length lists in the TLV factory:
- known length of the format `list[ LIST_SIZE ]`
- unknown length of the format `list[ ]`

From the programmer point of view, these are almost identical. The main
difference is that known length lists have the length variable part of
the TLV.
An Unknown length list has to be the last member in the TLV so its length
can be calculated using the TLV length field.

Variable length lists with known length (`list [ LIST_SIZE ]`) also
needs to be the last member of the TLV class due to insufficient support
in the TLV factory for variable length lists.
The reason for that is because all of the TLV class members are written
directly on the buffer (the class itself holds pointers to each member
on the buffer), so allocating a new list element for a list which is not
the last requires an update to all of the following members pointers,
and moving all data to the new "END" of the list.

This limitation is currently preventing us from properly implementing
some types of TLVs, such as M1 and M2 WSC which have variable length
strings such as 'manufacturer', 'device_name', etc.

This commit adds full support for variable length lists as follows:
Variable length members have an `alloc` method which is used to allocate
room on the buffer for the newly created list member.
On every such `alloc` function, tlvf.py will now also add a `memmove`
call to move all the remainder bytes following the current list "END" to
the new list "END". The number of bytes to move is simply calculated
using `getBuffRemainingBytes(void *start)` new API which was added for
this in a previous commit.
In addition, we also adds a "marker" (special comment used by the TLV
factory to mark locations to which we add new code when adding following
class members) to mark the location in the `alloc` function to which to
add code for each following member to update its pointer.

In case the list elements are of class type, for example:
```
cOuter:
  list_length:
    _type: uint16_t
    _length_var: True
  list:
    _type: cInner
    _length: [ list_length ]

cInner:
  list_length:
    _type: uint16_t
    _length_var: True
  list:
    _type: cInner
    _length: [ list_length ]
```

We also update the create method with a similar `memmove` since the
create method adds the class on the buffer with the initial size, before
the allocation method is called.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: update Autogenerated (variable length list)

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: align M1 and M2 to WSC specification

The EasyMesh Multi-AP specification defines the autoconfig flow which
uses WiFi Simple Setup (WSC) specification v2.0.6 M1 and M2 messages for
agent and controller fAPs and bAPs configuration.

The WSC specification defines the size of each parameter in the M1 and
M2 TLVs. Some of the attributes have fixed size, while others, the
stging attributes - 'manufacturer', 'model_name', 'device_name`,
`serial_number` and `model_number` - all are defined as variable length
lists of type CHAR as follows:
```
manufacturer_length:
  _type: uint16_t
  _length_var: True
manufacturer:
  _type: char
  _length: [ manufacturer_length ]
```

Due to the previous lack of support in the TLV factory for variable
length lists, these were defined with fixed length, which is not aligned
with the specification and will cause prplMesh to fail in certification
if the string sizes are not the masimum allowed by the WSC spec as we
defined them in the `WSC_attributes.yaml` file.
The outcome would be, in a very high probability, that prplMesh will not
pass certification and will not be able to work with other vendors which
implemented this correctly.

Since the previous commits added variable length lists support, this
commit updates the M1 and M2 TLVs to use variable length lists instead
of fixed sized strings, thus aligning to the WSC specification.

In addition, this commit also fixes the following:
- aligns the vendor extension attribute structure to the Multi-AP spec
- Move vendor extension definitions to TLVF defined in eWscVendorExt.yaml
- Removes sM1 and sM2 yamls since now we are using tlvWscM1 and tlvWscM2

Note - TLVF still lacks support for defining a variable length list if
the list member type is a class without an internal variable length
list, therefore all the variable length strings defined in the WSC
specification are moved to the TLVs themeselves.
This actually makes the usage cleaner, so not a high priority issue.
Issue is tracked in
https://github.com/prplfoundation/prplMesh/issues/115

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Update Autogenerated for WSC M1 and M2

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -445,12 +445,22 @@ def generateObject(self, obj_meta, dict_value):
                 param_type = param_meta.type
                 param_type_info = param_meta.type_info
+                is_enum_type = False
                 if param_type and param_type.startswith("e"):
```
this line should be replaced with:
`if TypeInfo(param_type).type == TypeInfo.ENUM`
That provides you the functionality you implemented in the following lines.

morantr reviewed (changes requested):
memmove() is banned function. according to: 
https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
Use instead std::move()
http://www.cplusplus.com/reference/algorithm/move/

morantr reviewed (dismissed):
What about all the class members pointers that are allocated after the list?
I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.

Commit:
tlvf: variable list of classes support

The previous commits added support for variable length lists but were
missing an update to the parameters pointers which come after the list
in case of classes (list type is class). This commit fixes that but due
to the TLVF design, create-(optional internal alloc)-add should be
atomic in that no changes to the buffer should be made by the user
between the create and add calls, since the member pointers may be
invalid at that time - See below for explanation.

Consider the following yaml:

cOuter:
  _type: class
  list1_length: uint16_t
  list1:
    _type: cInner
    _length: [ list1_length ]
  var1: uint32

Similarly `cInner` is a class with variable length list.
The issue is with var1's pointer which currently is not updated
correctly:

tlv = cmdu.addClass<cOuter>();  // list1 length is 0, m_var1 == m_list1

entry = create_list1();		// list1 length == sizeof(cInner), m_var1 == m_list1
				// so it point to an invalid location in the buffer and can't be
		    	        // used! but the actual value in the buffer already moved.
				// Only accessing the pointer at this time is problematic.

tlv.entry->alloc(1); 		// allocate internal list - no change to m_var1 which is
				// still invalid, values are memmoved correctly.

tlv.add_list(entry); 		// m_var1 and m_list1 are updated correctly.
				// The TLV values can be altered.

A drawing will help explain the issue:

Initial state:
 ____________________________
|_____|____________data______|
     / \
 var1   list1_end

create:
 ____________________________
|_____|XXX|________data______|
     /     \
 var1*      list1_end		// var1 still points to the start of the list and is invalid
				// The user MUST not access (especially change) its value.

alloc:

 ____________________________
|_____|XXX|YYY|____data______|
     /         \
 var1*          list1_end	// var1 still points to the start of the list and is invalid

internal alloc:

 ____________________________
|_____|XXX|YYY|ZZ|___data____|
     /            \
 var1*             list1_end    // var1 still points to the start of the list and is invalid

add:

 ____________________________
|_____|XXX|YYY|ZZ|_|___data__|
                  \
                   var1         // var1 is now valid

See yaml/tlvf/test/tlvVarList.yaml,
AutoGenerated/include/tlvf/test/tlvVarList.h and
Autogenerated/src/tlvf/test/tlvVarList.cpp for an example of the newly
generated files to reproduce this scenario.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

tomereli commented:
> memmove() is banned function. according to:
> https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
> Use instead std::move()
> http://www.cplusplus.com/reference/algorithm/move/

C memmove is banned - C++ std::memmove is fine (basically all of the STL APIs pass SDL)

tomereli commented:
> What about all the class members pointers that are allocated after the list?
> I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.

They are handled in the alloc function - read the commit description - when we build the alloc function in the TLVF, we add the `std::memmove` and add a special comment, "marker", that is later used when we are adding the parameters that come after the list. For each such parameter we add code using the marker to update the pointer before the `std::memmove`.

I have a bug which @LiorAmram pointed out - when the list element type is a class, we don't do that - if this is what you refer to then nice catch! I'll fix it in a followup commit.

morantr commented:
> > memmove() is banned function. according to:
> > https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
> > Use instead std::move()
> > http://www.cplusplus.com/reference/algorithm/move/
> 
> C memmove is banned - C++ std::memmove is fine (basically all of the STL APIs pass SDL)

According to here:
https://stackoverflow.com/questions/8380805/difference-between-string-h-and-cstring
it is said that:
"Table 55 describes the header <cstring>.
The contents are the same as the Standard C library header , with the change to memchr() specified in 21.7."

So basically it is the same C function with small change. I still think it is better to use std::move instead, it is more cpp way.

morantr commented:
> > What about all the class members pointers that are allocated after the list?
> > I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.
> 
> They are handled in the alloc function - read the commit description - when we build the alloc function in the TLVF, we add the `std::memmove` and add a special comment, "marker", that is later used when we are adding the parameters that come after the list. For each such parameter we add code using the marker to update the pointer before the `std::memmove`.
> 
> I have a bug which @LiorAmram pointed out - when the list element type is a class, we don't do that - if this is what you refer to then nice catch! I'll fix it in a followup commit.

Well, I looked at the generated code of list class type and didn't see that the member pointers are moved. So yes I meant that.


.

LiorAmram commented:
This [issue](https://github.com/prplfoundation/prplMesh/issues/115) should be addressed in the future.
For now it's OK to keep it as a working standard.

Commit:
tlvf: fix bug in swapping simple lists

swap function for lists was missing swap_prefix, so add that.
Fix can be seen in th eautogenerated test yaml.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: update autogenerated for tlvOperatingChannelReport

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: var length lists - use std::copy_n

Change to use std::copy_n instead of std::memmove due to SDK
requirements.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Update Autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: forcing allocation order in multiple lists

Fix issue https://github.com/prplfoundation/prplMesh/issues/116

There is a gap in tlvf in a case there is a simple list followed by a
nested list.
for example:

tlvSample:
    _type: class
    _is_tlv_class: true
    type: eTlvType
    length: uint16_t

    simple_list_length:
        _type: uint16_t
        _length_var: true
    simple_list:
        _type: uint16_t
        _length: [simple_list_length]

    complex_list_length:
        _type: uint16_t
        _length_var: true
    complex_list:
        _type: cInner
        _length: [complex_list_length]

cInner:
    _type: class
    inner_list_length:
        _type: uint16_t
        _length_var: true
    inner_list:
        _type: uint16_t
        _length: [inner_list_length]
    inner_param: uint32_t

In this case, allocation of simple list won't be aware of the the
inner_param (notice there are multiple instances of it, because
complex_list can contain multiple entries), and therefore won't update
its pointer, which will cause data corruption.

Updating all the following pointers is very complex, and may result in a
very complex generated code.
My suggestion is 'locking' the allocation, forcing the user to allocate
lists' entries by order.
Each class will have a integer lock_order_counter which will be
initialized to 0.
Each list/inner class will have its own list_index (internally, can be
managed in python, and generated as hard coded 'magic numbers' in the
c++ code. Doesn't require additional variables).
When calling alloc/create - the method will check the
lock_order_counter:
if lock_order_counter > list_index - return false.
if lock_order_counter <= list_index - set lock_order_counter =
list_index and continue.

That way, the user will be forced to to work on the lists in order, and
we can avoid corrupted data.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
Update Autogenerated files

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -726,6 +731,8 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
             if is_var_len:
                 # variable length list support - swap length value for calculating list size
                 length_type = obj_meta.children_types[param_length]
+                param_meta.list_index = obj_meta.list_index
```
This mechanism should also be added to dynamic lists (is_dynamic_len - unknown list length).

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -874,6 +881,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
+                    lines_cpp.append( "%sif (m_%s__ > %s) {" %(self.getIndentation(1), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
+                    lines_cpp.append( "%sTLVF_LOG(ERROR) << \"Out of order allocation for variable length list %s, abort!\";" %(self.getIndentation(2), param_name) )
+                    lines_cpp.append( "%sreturn nullptr;" %self.getIndentation(2))
+                    lines_cpp.append( "%s} else {" %self.getIndentation(1))
```
'else' section is unnecessary, because of the return statement.

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -874,6 +881,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
+                    lines_cpp.append( "%sTLVF_LOG(ERROR) << \"Out of order allocation for variable length list %s, abort!\";" %(self.getIndentation(2), param_name) )
+                    lines_cpp.append( "%sreturn nullptr;" %self.getIndentation(2))
+                    lines_cpp.append( "%s} else {" %self.getIndentation(1))
+                    lines_cpp.append( "%sm_%s__ = %s;" %(self.getIndentation(2), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
```
Might be better to first make all sanity tests (count == 0 and available space on buffer) and only then update the counter.

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -926,6 +939,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
                 else: #simple list
                     lines_h.append( "bool alloc_%s(size_t count = 1);" % (param_name) )
                     lines_cpp.append( "bool %s::alloc_%s(size_t count) {" % (obj_meta.name, param_name) )
+                    lines_cpp.append( "%sif (m_%s__ > %s) {;" %(self.getIndentation(1), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
```
remove semicolon at the end of this line

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -926,6 +939,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
+                    lines_cpp.append( "%sif (m_%s__ > %s) {;" %(self.getIndentation(1), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
+                    lines_cpp.append( "%sTLVF_LOG(ERROR) << \"Out of order allocation for variable length list %s, abort!\";" %(self.getIndentation(2), param_name) )
+                    lines_cpp.append( "%sreturn false;" %self.getIndentation(2))
+                    lines_cpp.append( "%s} else {" %self.getIndentation(1))
```
same - remove else

LiorAmram commented on framework/tlvf/tlvf.py:
```diff
@@ -926,6 +939,12 @@ def addClassParam(self, obj_meta, param_name, param_type, param_type_info, param
+                    lines_cpp.append( "%sTLVF_LOG(ERROR) << \"Out of order allocation for variable length list %s, abort!\";" %(self.getIndentation(2), param_name) )
+                    lines_cpp.append( "%sreturn false;" %self.getIndentation(2))
+                    lines_cpp.append( "%s} else {" %self.getIndentation(1))
+                    lines_cpp.append( "%sm_%s__ = %s;" %(self.getIndentation(2), self.MEMBER_LOCK_ORDER_COUNTER, param_meta.list_index) )
```
same

tomereli commented:
> memmove() is banned function. according to:
> https://github.com/intel/safestringlib/wiki/SDL-List-of-Banned-Functions
> Use instead std::move()
> http://www.cplusplus.com/reference/algorithm/move/

Done

tomereli commented:
> > > What about all the class members pointers that are allocated after the list?
> > > I mean, they are set to a specific location on the buffer on the init function. And they are still pointing to the same place after you moved the memory. Changing them will still create a data override.
> > 
> > 
> > They are handled in the alloc function - read the commit description - when we build the alloc function in the TLVF, we add the `std::memmove` and add a special comment, "marker", that is later used when we are adding the parameters that come after the list. For each such parameter we add code using the marker to update the pointer before the `std::memmove`.
> > I have a bug which @LiorAmram pointed out - when the list element type is a class, we don't do that - if this is what you refer to then nice catch! I'll fix it in a followup commit.
> 
> Well, I looked at the generated code of list class type and didn't see that the member pointers are moved. So yes I meant that.
> 
> .

It all works fine now. Generated code does update the flag:
```
bool cPreferenceOperatingClasses::alloc_channel_list(size_t count) {
    if (m_lock_order_counter__ > 0) {;
        TLVF_LOG(ERROR) << "Out of order allocation for variable length list channel_list, abort!";
        return false;
    } else {
        m_lock_order_counter__ = 0;
    }
    if (count == 0) {
        TLVF_LOG(WARNING) << "can't allocate 0 bytes";
        return false;
    }
    size_t len = sizeof(uint8_t) * count;
    if(getBuffRemainingBytes() < len )  {
        TLVF_LOG(ERROR) << "Not enough available space on buffer - can't allocate";
        return false;
    }
    if (!m_parse__) {
        uint8_t *src = (uint8_t *)m_channel_list;
        uint8_t *dst = (uint8_t *)m_channel_list + len;
        size_t move_length = getBuffRemainingBytes(src) - len;
        std::copy_n(src, move_length, dst);
    }
    m_flags = (sFlags *)((uint8_t *)(m_flags) + len); // <----- update m_flags pointer
    m_channel_list_idx__ += count;
    *m_channel_list_length += count;
    m_buff_ptr__ += len;
    return true;
}
```

I also triggered channel selection from BML and changed the flags reason code to `OPERATION_WOULD_PREVENT_BACKHAUL_OPERATION_USING_SHARED_RADIO` to make sure it works, and it does:

![image](https://user-images.githubusercontent.com/13386849/61537886-e3eb7a80-aa40-11e9-8411-fa379c8c2014.png)

Commit:
agent: fix order of list creatin in M1

Due to the previous commit which added protection against adding lists
out of order, the agent WSC M1 message failed to add the serial number
attribute in the WSC M1 TLV since it was done out of order.
This commit changes the order of variables creation and fixes the issue.

Actually this protection is not really needed in this case since only if
the list member type is another complex list, a buffer corruption may
occur.
But adapting the TLVF python script to support full out of order
assignments is too complex at this stage of the project.
The alternative of removing the protection is also bad since it is far
better to have a warning print where its not really necessary and
failing the list allocation than having corrupt buffer.

While at it - also add validity checks for each assignment and return
false on error, in both son_master_thread and son_slave_thread.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: Fix complex list potential bug in create/add

If a complex list is allocated, the initial size is moved but we won't
update the member pointers until add() API is called.
This is problematic since the user might call create, then update the
values of the returned class pointer, then before calling add() also
modify following class members thus overriding the changes done.

This commit fixes this corner case by updating the members pointers both
on create() and add().

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>

Commit:
tlvf: add README.md

Port TLVF documentation from word to markdown and update variable length
lists section.

Signed-off-by: Tomer Eliyahu <tomer.b.eliyahu@intel.com>
